File: lightroom_tagger/core/database/embeddings.py
"""CLIP embedding DB helpers."""

import sqlite3


def _sort_catalog_key_rows_newest_first(rows: list[sqlite3.Row]) -> list[str]:
    """Return keys sorted by ``date_taken`` desc, then key desc."""
    keyed: list[tuple[str, str]] = []
    for row in rows:
        key = str(row["key"])
        date_taken = str(row["date_taken"] or "")
        keyed.append((key, date_taken))
    keyed.sort(key=lambda item: (item[1], item[0]), reverse=True)
    return [key for key, _ in keyed]


def upsert_image_clip_embedding(
    conn: sqlite3.Connection, image_key: str, embedding_blob: bytes
) -> None:
    """Replace vec0 row for ``image_key``. Call inside :func:`library_write` only."""
    conn.execute("DELETE FROM image_clip_embeddings WHERE image_key = ?", (image_key,))
    conn.execute(
        "INSERT INTO image_clip_embeddings(embedding, image_key) VALUES (?, ?)",
        (embedding_blob, image_key),
    )


def _list_catalog_keys_clip_embed_sql_params(
    *,
    months: int | None,
    year: str | None,
    min_rating: int | None,
) -> tuple[str, tuple]:
    parts: list[str] = [
        "i.filepath IS NOT NULL AND TRIM(COALESCE(i.filepath, '')) != ''",
    ]
    params: list = []
    if months is not None:
        parts.append("i.date_taken >= date('now', ?)")
        params.append(f"-{months} months")
    if year is not None:
        parts.append("strftime('%Y', i.date_taken) = ?")
        params.append(year)
    if min_rating is not None:
        parts.append("i.rating >= ?")
        params.append(min_rating)
    where = " AND ".join(parts)
    sql = f"""
        SELECT i.key AS key, i.date_taken AS date_taken
        FROM images i
        WHERE {where}
        ORDER BY i.key ASC
    """
    return sql, tuple(params)


def list_catalog_keys_needing_clip_embedding(
    conn: sqlite3.Connection,
    *,
    months: int | None,
    year: str | None,
    min_rating: int | None,
) -> list[str]:
    """Catalog keys with a usable file path in the date/rating window, missing CLIP vec0 rows."""
    sql, params = _list_catalog_keys_clip_embed_sql_params(
        months=months,
        year=year,
        min_rating=min_rating,
    )
    rows = conn.execute(sql, params).fetchall()
    embedded_keys = {
        str(r["image_key"])
        for r in conn.execute("SELECT image_key FROM image_clip_embeddings").fetchall()
    }
    filtered_rows = [r for r in rows if str(r["key"]) not in embedded_keys]
    return _sort_catalog_key_rows_newest_first(filtered_rows)


def list_catalog_keys_for_clip_embed_force(
    conn: sqlite3.Connection,
    *,
    months: int | None,
    year: str | None,
    min_rating: int | None,
) -> list[str]:
    """All catalog keys with a usable file path in the window, including keys in ``image_clip_embeddings``."""
    sql, params = _list_catalog_keys_clip_embed_sql_params(
        months=months,
        year=year,
        min_rating=min_rating,
    )
    rows = conn.execute(sql, params).fetchall()
    return _sort_catalog_key_rows_newest_first(rows)

File: apps/visualizer/backend/src/db/library/embeddings.ts
/**
 * Writing and selecting CLIP embedding rows.
 *
 * `image_clip_embeddings` is a sqlite-vec `vec0` virtual table, which has no
 * UPDATE path for the vector column — a re-embed is a delete followed by an
 * insert, and both must happen inside one `libraryWrite` so a crash between them
 * cannot leave the image with no vector at all.
 */
import type { Db } from '../connection.js';

/** Replace the vector for `imageKey`. Call inside `libraryWrite`. */
export function upsertImageClipEmbedding(db: Db, imageKey: string, embeddingBlob: Buffer): void {
  db.prepare('DELETE FROM image_clip_embeddings WHERE image_key = ?').run(imageKey);
  db.prepare('INSERT INTO image_clip_embeddings(embedding, image_key) VALUES (?, ?)').run(
    embeddingBlob,
    imageKey,
  );
}

export interface ClipEmbedWindow {
  months: number | null;
  year: string | null;
  minRating: number | null;
}

/**
 * Newest first, undated last — the order the embed job works through the catalog.
 *
 * `COALESCE(date_taken, '')` orders undated rows last under SQLite's default collation.
 */
function windowedCatalogKeys(db: Db, window: ClipEmbedWindow): string[] {
  const conditions = ["i.filepath IS NOT NULL AND TRIM(COALESCE(i.filepath, '')) != ''"];
  const params: unknown[] = [];

  if (window.months !== null) {
    conditions.push("i.date_taken >= date('now', ?)");
    params.push(`-${window.months} months`);
  }
  if (window.year !== null) {
    conditions.push("strftime('%Y', i.date_taken) = ?");
    params.push(window.year);
  }
  if (window.minRating !== null) {
    conditions.push('i.rating >= ?');
    params.push(window.minRating);
  }

  const rows = db
    .prepare(
      `SELECT i.key AS key FROM images i WHERE ${conditions.join(' AND ')}` +
        " ORDER BY COALESCE(i.date_taken, '') DESC, i.key DESC",
    )
    .all(...(params as never[])) as { key: string }[];
  return rows.map((r) => r.key);
}

/** Catalog keys in the window that have no vector yet. */
export function listCatalogKeysNeedingClipEmbedding(db: Db, window: ClipEmbedWindow): string[] {
  const keys = windowedCatalogKeys(db, window);
  // Read the embedded keys in one pass rather than joining: `vec0` is a virtual
  // table, and a correlated subquery against one is not something to rely on.
  const embedded = new Set(
    (db.prepare('SELECT image_key FROM image_clip_embeddings').all() as { image_key: string }[]).map(
      (r) => String(r.image_key),
    ),
  );
  return keys.filter((k) => !embedded.has(k));
}

/** Every catalog key in the window, including ones already embedded. */
export function listCatalogKeysForClipEmbedForce(db: Db, window: ClipEmbedWindow): string[] {
  return windowedCatalogKeys(db, window);
}
